### PR TITLE
Fixed and added Tables unit tests.

### PR DESCRIPTION
**Pull request by vcolin7**
Also added recordings.

I had to modify some generated models to fit the XML shape the Storage service expects, for some reason the swagger was not enough for autorest to generate things properly. I will investigate further. It might come It may come down to some setting or adding transforms in the worst case. 

Opened the following issue to track this: #22195.

### CODE REVIEW

File: sdk/tables/azure-data-tables/pom.xml [hunk 1]
```diff
@@ -32,10 +32,6 @@ Licensed under the MIT License.
     <tag>HEAD</tag>
   </scm>
 
-  <properties>
-    <jacoco.skip.coverage.check>true</jacoco.skip.coverage.check>
```
Comment (srnagar) [inline sdk/tables/azure-data-tables/pom.xml:36]: 👍

File: sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableServiceAsyncClient.java [hunk 2]
```diff
@@ -142,6 +144,15 @@ ClientLogger getLogger() {
         return this.logger;
     }
 
+    /**
+     * Gets this client's {@link SerializerAdapter}.
+     *
+     * @return This client's {@link SerializerAdapter}.
+     */
+    SerializerAdapter getSerializerAdapter() {
```
Comment (srnagar) [inline sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableServiceAsyncClient.java:152]: Why is this required? This must be an implementation detail and shouldn't be used in tests.

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableServiceAsyncClientTest.java [hunk 3]
```diff
@@ -419,25 +427,84 @@ public void canUseSasTokenToCreateValidTableClient() {
...
+            secondaryHostJoiner.add(hostParts[i]);
+        }
 
-        assertNotNull(statistics);
-        assertNotNull(statistics.getGeoReplication());
-        assertNotNull(statistics.getGeoReplication().getStatus());
-        assertNotNull(statistics.getGeoReplication().getLastSyncTime());
+        String secondaryEndpoint = primaryEndpoint.getScheme() + "://" + secondaryHostJoiner;
+
+        TableServiceAsyncClient secondaryClient = new TableServiceAsyncClient(serviceClient.getHttpPipeline(),
```
Comment (srnagar) [inline sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableServiceAsyncClientTest.java:497]: Why is this client created using the constructor and not the builder?

Comment (vcolin7) [reply sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableServiceAsyncClientTest.java:497]: Felt it was quicker. I can change it to use the builder so we avoid adding `getSerializerAdapter()` to the async client.

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java [hunk 4]
```diff
@@ -0,0 +1,1057 @@
...
+                    + "&erk=endRowKey"
+                    + "&sip=a-b"
+                    + "&spr=https%2Chttp"
+                    + "&sig="
+            )
+        );
+    }
+
+    @Test
+    @Tag("SAS")
```
Comment (srnagar) [inline sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java:941]: What's the purpose of this tag? Is it used anywhere to selectively run tests?

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableServiceAsyncClientTest.java [hunk 5]
```diff
@@ -419,25 +427,87 @@ public void canUseSasTokenToCreateValidTableClient() {
...
-        assertNotNull(statistics);
-        assertNotNull(statistics.getGeoReplication());
-        assertNotNull(statistics.getGeoReplication().getStatus());
-        assertNotNull(statistics.getGeoReplication().getLastSyncTime());
+        String secondaryEndpoint = primaryEndpoint.getScheme() + "://" + secondaryHostJoiner;
+
+        TableServiceAsyncClient secondaryClient = new TableServiceClientBuilder()
+            .endpoint(secondaryEndpoint)
+            .serviceVersion(serviceClient.getServiceVersion())
+            .pipeline(serviceClient.getHttpPipeline())
```
Comment (srnagar) [inline sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableServiceAsyncClientTest.java:500]: A new pipeline is created if it's not provided in the builder. So, this isn't required.

File: sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableServiceClient.java [hunk 6]
```diff
@@ -73,6 +74,15 @@ public TableServiceVersion getServiceVersion() {
         return client.getServiceVersion();
     }
 
+    /**
+     * Gets the {@link HttpPipeline} powering this client.
+     *
+     * @return This client's {@link HttpPipeline}.
+     */
+    HttpPipeline getHttpPipeline() {
```
Comment (srnagar) [inline sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableServiceClient.java:82]: This API is not required.

**Review by srnagar — APPROVED** [review]

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableServiceAsyncClientTest.java (hunk 5 above)
Comment (vcolin7) [reply sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableServiceAsyncClientTest.java:500]: I thought it would be better to use the same pipeline from the first client to use the same configuration. I think can write things better to make it so that we can get those settings in the test class itself.

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java (hunk 4 above)
Comment (vcolin7) [reply sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java:941]: Not at the moment. Other tests had these tags so I thought it was ok to follow the pattern in case we wanted to use them in the future. I can remove them for now and things shouldn't break or anything.

File: sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/implementation/models/TableServiceStats.java [hunk 7]
```diff
@@ -6,8 +6,10 @@
 
 import com.azure.core.annotation.Fluent;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement;
 
 /** Stats for the service. */
+@JacksonXmlRootElement(localName = "StorageServiceStats")
 @Fluent
 public final class TableServiceStats {
```
Comment (alzimmermsft) [inline sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/implementation/models/TableServiceStats.java:14]: Was there a reason to use `Stats` instead of `Statistics` used in the other Storage SDKs?

File: sdk/tables/azure-data-tables/src/samples/java/com/azure/data/tables/TransactionalBatchOperation.java [hunk 8]
```diff
@@ -39,33 +39,33 @@ public static void main(String[] args) {
             .addProperty("Brand", "Crayola")
             .addProperty("Color", "Red");
         transactionActions.add(new TableTransactionAction(TableTransactionActionType.CREATE, firstEntity));
-        System.out.printf("Added create operation to transactional batch for entity with partition key: %s, and row "
-            + "key: %s\n", partitionKey, firstEntityRowKey);
+        System.out.printf("Added create action for entity with partition key: %s, and row key: %s\n", partitionKey,
```
Comment (alzimmermsft) [inline sdk/tables/azure-data-tables/src/samples/java/com/azure/data/tables/TransactionalBatchOperation.java:42]: nit: `%n` not `\n`

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java [hunk 9]
```diff
@@ -0,0 +1,1044 @@
...
+            .connectionString(connectionString)
+            .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BODY_AND_HEADERS))
+            .tableName(tableName);
+
+        if (interceptorManager.isPlaybackMode()) {
+            playbackClient = interceptorManager.getPlaybackClient();
+
+            builder.httpClient(playbackClient);
+        } else {
+            builder.httpClient(HttpClient.createDefault());
```
Comment (alzimmermsft) [inline sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java:74]: Should use a static for `HttpClient`, right now this is creating a new `HttpClient` per test which is a costly operation. This doesn't need to be a change in this PR.

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java [hunk 10]
```diff
@@ -0,0 +1,1044 @@
...
+            if (!interceptorManager.isLiveMode()) {
+                recordPolicy = interceptorManager.getRecordPolicy();
+
+                builder.addPolicy(recordPolicy);
+            }
+
+            builder.addPolicy(new RetryPolicy(new ExponentialBackoff(6, Duration.ofMillis(1500),
```
Comment (alzimmermsft) [inline sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java:81]: Any reason for this custom retry policy?

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java [hunk 11]
```diff
@@ -0,0 +1,1044 @@
...
+
+        if (interceptorManager.isPlaybackMode()) {
+            builder.httpClient(playbackClient);
+        } else {
+            builder.httpClient(HttpClient.createDefault());
+            if (!interceptorManager.isLiveMode()) {
+                builder.addPolicy(recordPolicy);
+            }
+            builder.addPolicy(new RetryPolicy(new ExponentialBackoff(6, Duration.ofMillis(1500),
+                Duration.ofSeconds(100))));
```
Comment (alzimmermsft) [inline sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java:107]: This logic seems very duplicative, could this be made into a separate method

File: sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/implementation/models/TableServiceStats.java (hunk 7 above)
Comment (vcolin7) [reply sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/implementation/models/TableServiceStats.java:14]: It's a generated class so it looks like this is how the service folks specified the model should be named.

File: sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java (hunk 10 above)
Comment (vcolin7) [reply sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java:81]: Not really, I can look into it a bit.

Comment (vcolin7) [reply sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TableClientTest.java:81]: I think it's there just to be super safe in live tests, but I don't thinks it's necessarily required. I'll get rid of it for now.
